### PR TITLE
Add integration test for NVR edge cases 

### PR DESCRIPTION
**Pull request by vaibhavdaren**
Add test to verify file artifact provider correctly handles NVR edge cases. 
This is continued from discussion on the PR for file provider:   https://github.com/teemtee/tmt/pull/4510#issuecomment-3848015990

Related: #4546

#### Why it falls back on Rawhide

The `tmt-artifact-shared` repository is created with **priority 50**. However, because packages are currently requested by **name** (e.g., `jq`) rather than by the exact **NVR** (Name-Version-Release), `dnf` has too much flexibility.

When the downloaded RPM (e.g., `jq-1.8.1-2.fc44.x86_64`) has dependencies that are incompatible with the Rawhide host, `dnf` determines it cannot install that specific file from the artifact repo. Because the request was simply for `jq`, it silently resolves the package from `updates` or `fedora` instead to satisfy the request.

#### Proposed Path Forward

To ensure the integrity of the test environment, we should implement the following:

1. **Strict NVR Enforcement:** Modify the installation logic to request the exact NVR. This forces `dnf` to either install the specific artifact or fail if dependencies aren't met.
2. **Post-Installation Verification:** After a package is installed, verify the source repository. If the package was pulled from a system repo instead of `tmt-artifact-shared` (or a repo defined by the provider), the process should raise an error with detailed logs.

#### Why it works with dependencies included

When we download using `--resolve --alldeps`, the artifact repository contains the full dependency chain. This allows `dnf` to satisfy the entire transaction locally from `tmt-artifact-shared` without needing to query default system repositories. You can see this passing in the test case:
[/tests/prepare/artifact/nvr-edge-cases/package-with-dependency](/tests/prepare/artifact/nvr-edge-cases/package-with-dependency)

### CODE REVIEW

File: tests/prepare/artifact/nvr-edge-cases/package-with-dependency/data/test.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+rpm -q jq
+
+# Verify jq came from the tmt-artifact-shared repository
+dnf info --installed jq | grep -Eq "From repo(sitory)?\s*:\s*tmt-artifact-shared"
+
+# Verify nano is installed (from directory with multiple RPMs)
+rpm -q nano
+
+# Verify nano came from the tmt-artifact-shared repository
+dnf info --installed nano | grep -Eq "From repo(sitory)?\s*:\s*tmt-artifact-shared"
```
[the remark this answers is not in the document]
Comment (vaibhavdaren) [reply tests/prepare/artifact/nvr-edge-cases/package-with-dependency/data/test.sh]: The test is supposed to fail in fedora-rawhide. Waiting for the failed run. So, I can get the required logs.  I will make the suggested changes post that.

**Comment by vaibhavdaren** [comment]
**Note**🥇: Isolated the failing test ^_^ in [testing-farm:fedora-rawhide-x86_64:full](https://github.com/teemtee/tmt/runs/62868640050)    
https://artifacts.dev.testing-farm.io/f4e74ca8-7d1a-43f6-961c-e62e586f157b

**Comment by LecrisUT** [comment]
Ok, did some local testing to see how this behaves, indeed the priority takes precedence over all this. @vaibhavdaren @AthreyVinay  Do you have a document to put the test cases and behavior?

**Comment by vaibhavdaren** [comment]
> Ok, did some local testing to see how this behaves, indeed the priority takes precedence over all this. @vaibhavdaren @AthreyVinay Do you have a document to put the test cases and behavior?

We dont have a file, I think we can use https://github.com/teemtee/tmt/issues/4546 to document our findings.

**Comment by vaibhavdaren** [comment]
When dependency are present in the shared artifact dir the tests pass on rawhide: https://artifacts.dev.testing-farm.io/361bd67e-9794-4863-8d91-f2152506bc83

**Comment by LecrisUT** [comment]
> When dependency are present in the shared artifact dir the tests pass on rawhide: https://artifacts.dev.testing-farm.io/361bd67e-9794-4863-8d91-f2152506bc83

Can you elaborate? Basically the fact that `jq` built F43 failed is expected, but the fact that it fallbacked to system is unexpected and we **must** prevent that in the final design, i.e. the installation **must** fail with that exact packages requested. If you make it install `jq-1.8.1-2.fc44.x86_64` instead of `jq` it will fail as we expect it to. I have put some more details of observations in the google docs, but I will move it to #4546, let's move the discussion there.

**Comment by vaibhavdaren** [comment]
### Analysis: Unexpected Fallback to System Repositories

It sounds like we’ve pinpointed the exact point of failure. `dnf` is doing what it’s designed to do : find *a* working solution when what we actually want is for it to **fail fast** if our specific artifact is incompatible.
We are on the same page: while the failure of the `jq` build on F43 is expected, the silent fallback to system repositories is a behavior we must prevent in the final design.

#### Why it falls back on Rawhide

The `tmt-artifact-shared` repository is created with **priority 50**. However, because packages are currently requested by **name** (e.g., `jq`) rather than by the exact **NVR** (Name-Version-Release), `dnf` has too much flexibility.

When the downloaded RPM (e.g., `jq-1.8.1-2.fc44.x86_64`) has dependencies that are incompatible with the Rawhide host, `dnf` determines it cannot install that specific file from the artifact repo. Because the request was simply for `jq`, it silently resolves the package from `updates` or `fedora` instead to satisfy the request.

#### Proposed Path Forward

To ensure the integrity of the test environment, we should implement the following:

1. **Strict NVR Enforcement:** Modify the installation logic to request the exact NVR. This forces `dnf` to either install the specific artifact or fail if dependencies aren't met.
2. **Post-Installation Verification:** After a package is installed, verify the source repository. If the package was pulled from a system repo instead of `tmt-artifact-shared` (or a repo defined by the provider), the process should raise an error with detailed logs.

#### Why it works with dependencies included

When we download using `--resolve --alldeps`, the artifact repository contains the full dependency chain. This allows `dnf` to satisfy the entire transaction locally from `tmt-artifact-shared` without needing to query default system repositories. You can see this passing in the test case:
`/tests/prepare/artifact/nvr-edge-cases/package-with-dependency`

**Comment by LecrisUT** [comment]
> #### Why it falls back on Rawhide
> 
> The `tmt-artifact-shared` repository is created with **priority 50**. However, because packages are currently requested by **name** (e.g., `jq`) rather than by the exact **NVR** (Name-Version-Release), `dnf` has too much flexibility.

Note that dnf has a flag `--from-repo`, from my testing it worked as intended, but not very well with name-only, mostly with NVR. Don't remember if I tried it with the failing package as well, so try that.

> 1. **Strict NVR Enforcement:** Modify the installation logic to request the exact NVR. This forces `dnf` to either install the specific artifact or fail if dependencies aren't met.

Edge case to consider here is what if 2 repos provide the same NVR. My guess is that `--from-repo` would take care of this issue, but it's best if it is confirmed.

> Why it works with dependencies included

This should not be relevant, and should be avoided. The case here is a perfect example of why we should not be doing it as it breaks what is expected to have failed.

**Comment by AthreyVinay** [comment]
>  My guess is that `--from-repo` would take care of this issue, but it's best if it is confirmed.

Would be interested to know the findings too.

File: tests/prepare/artifact/nvr-debug/data/build-repos.sh [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+REPO_LOW="$REPO_BASE/repo-low"
+BUILD_DIR="/tmp/rpm-build-nvr-test"
+
+dnf install -y rpm-build createrepo_c
+
+mkdir -p "$REPO_HIGH" "$REPO_LOW"
+mkdir -p "$BUILD_DIR"/{BUILD,RPMS,SOURCES,SPECS,SRPMS}
+
+# Spec for v1.0 (goes into the high-priority repo)
+cat > "$BUILD_DIR/SPECS/v1.spec" << 'SPECEOF'
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/data/build-repos.sh]: Please make thesr concrete files

File: tests/prepare/artifact/nvr-debug/data/build-repos.sh [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+PACKAGE="dummy-nvr-test"
+REPO_BASE="/tmp/nvr-test"
+REPO_HIGH="$REPO_BASE/repo-high"
+REPO_LOW="$REPO_BASE/repo-low"
+BUILD_DIR="/tmp/rpm-build-nvr-test"
+
+dnf install -y rpm-build createrepo_c
+
+mkdir -p "$REPO_HIGH" "$REPO_LOW"
+mkdir -p "$BUILD_DIR"/{BUILD,RPMS,SOURCES,SPECS,SRPMS}
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/data/build-repos.sh]: This folder structure is not necessary, the specs can be in arbitrary location

File: tests/prepare/artifact/nvr-debug/data/build-repos.sh [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+set -ex
+
+PACKAGE="dummy-nvr-test"
+REPO_BASE="/tmp/nvr-test"
+REPO_HIGH="$REPO_BASE/repo-high"
+REPO_LOW="$REPO_BASE/repo-low"
+BUILD_DIR="/tmp/rpm-build-nvr-test"
+
+dnf install -y rpm-build createrepo_c
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/data/build-repos.sh]: Building with mock is preferred. And splitting the srpm and binary rpm build would be useful since the srpm could be cached and reused across tested distros (and even across runs when that would be possible)

File: tests/prepare/artifact/nvr-debug/data/build-repos.sh [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+%files
+
+%changelog
+SPECEOF
+
+rpmbuild --define "_topdir $BUILD_DIR" -bb "$BUILD_DIR/SPECS/v1.spec"
+rpmbuild --define "_topdir $BUILD_DIR" -bb "$BUILD_DIR/SPECS/v2.spec"
+
+find "$BUILD_DIR/RPMS" -name "${PACKAGE}-1.0*.rpm" -exec cp {} "$REPO_HIGH/" \;
+find "$BUILD_DIR/RPMS" -name "${PACKAGE}-2.0*.rpm" -exec cp {} "$REPO_LOW/" \;
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/data/build-repos.sh]: The `find` is so unnecessary. Just put that glob directly in the `cp`

File: tests/prepare/artifact/nvr-debug/data/plan.fmf [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    discover:
+        how: fmf
+        filter: 'name: /tc12'
+    provision:
+        how: container
+    prepare:
+        - name: Build dummy RPMs and create local repos
+          how: shell
+          order: 40
+          script: ./build-repos.sh
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/data/plan.fmf]: Why is this run inside the plan?

File: tests/prepare/artifact/nvr-debug/data/tc11-high.repo [hunk 7]
```diff
@@ -0,0 +1,6 @@
+[test-high-priority]
+name=Test High Priority Repo (dummy-nvr-test v1.0)
+baseurl=file:///tmp/nvr-test/repo-high
+enabled=1
+gpgcheck=0
+priority=1
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/data/tc11-high.repo]: Let's avoid extreme numbers. `50` and `30` seem fine mid-range values

File: tests/prepare/artifact/nvr-debug/test.sh [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+                --provide repository-file:file://$(pwd)/tc11-low.repo" \
+            0 "TC 1.1: high-priority repo (v1.0) beats low-priority repo (v2.0)"
+    rlPhaseEnd
+
+    rlPhaseStartTest "TC 1.2 - Same priority falls back to NVR comparison"
+        rlRun "tmt run -i $run --scratch -vvv --all \
+            plan -n /tc12 \
+            provision -h $PROVISION_HOW --image $TEST_IMAGE_PREFIX/$image_name \
+            prepare --insert --how artifact \
+                --provide repository-file:file://$(pwd)/tc12-a.repo \
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/test.sh]: The usage of artifact provide seems arbitrary here if you do all of the building in the nested plan. If you move the build up then at least you can reuse the artifacts.

File: tests/prepare/artifact/nvr-debug/data/build-repos.sh (hunk 4 above)
Comment (vaibhavdaren) [reply tests/prepare/artifact/nvr-debug/data/build-repos.sh]: Done! I’ve optimized this by caching the SRPMs. They are now generated on the host logic inside `build-srpms/` and saved to /data. This ensures we're using a single, consistent source artifact for all downstream builds across different distributions.

File: tests/prepare/artifact/nvr-debug/rpm-data/build-srpms.sh [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+mkdir -p "$SRPM_DIR"
+
+mock --buildsrpm --spec "$SCRIPT_DIR/v1.spec" --sources "$SCRIPT_DIR/" \
+    --resultdir "$SRPM_DIR" --enable-network
+mock --buildsrpm --spec "$SCRIPT_DIR/v2.spec" --sources "$SCRIPT_DIR/" \
+    --resultdir "$SRPM_DIR" --enable-network
+
+# This is done to speed up the tests
+echo "SRPMs staged in $SRPM_DIR — commit them with:"
+echo "  git add data/srpms/*.src.rpm && git commit"
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/rpm-data/build-srpms.sh]: Put these in `nvr-debug/test.sh` and do not git commit these. We can figure out a better caching of these later (I will even start the first parts of that design now)

File: tests/prepare/artifact/nvr-debug/rpm-data/build-srpms.sh [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+# Run this whenever the spec files change, then commit data/srpms/.
+#
+# Usage:
+#   ./build_srpms.sh
+#   git add data/srpms/*.src.rpm
+#   git commit -m "Update committed SRPMs"
+
+set -ex
+
+SCRIPT_DIR="$(dirname "$(readlink -f "$0")")"
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/rpm-data/build-srpms.sh]: Unnecessarily complicated
```suggestion
SCRIPT_DIR="$(dirname $0)"
```
That's fine enough.

File: tests/prepare/artifact/nvr-debug/rpm-data/build-srpms.sh [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+SRPM_DIR="$SCRIPT_DIR/../data/srpms"
+
+mkdir -p "$SRPM_DIR"
+
+mock --buildsrpm --spec "$SCRIPT_DIR/v1.spec" --sources "$SCRIPT_DIR/" \
+    --resultdir "$SRPM_DIR" --enable-network
+mock --buildsrpm --spec "$SCRIPT_DIR/v2.spec" --sources "$SCRIPT_DIR/" \
+    --resultdir "$SRPM_DIR" --enable-network
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/rpm-data/build-srpms.sh]:
```suggestion
for spec in $SCRIPT_DIR/*.spec; do
  mock --buildsrpm --spec "$SCRIPT_DIR/v1.spec" --sources "$SCRIPT_DIR/" \
    --resultdir "$SRPM_DIR"
done
```
Don't see why we wouldn't build all srpms

File: tests/prepare/artifact/nvr-debug/data/plan.fmf [hunk 12]
```diff
@@ -0,0 +1,11 @@
+summary: NVR priority and version selection tests
+discover:
+    how: fmf
+provision:
+    how: container
+prepare:
+    - name: Build dummy RPMs and create local repos
+      how: shell
+      script: ./build-repos.sh
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/data/plan.fmf]:
```suggestion
    - name: Build dummy RPMs and create local repos
      order: 20
      how: shell
      script: ./build-repos.sh
```
Still not a fan of building these inside the inner tests, but at least run this before the artifacts

File: tests/prepare/artifact/nvr-debug/data/build-repos.sh [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+# by tmt as part of the normal repo sync.
+# Both test cases (tc11, tc12) share these repos within the same container.
+
+set -ex
+
+PACKAGE="dummy-nvr-test"
+BUILD_DIR="$(mktemp -d)"
+REPO_BASE="/tmp/nvr-test"
+
+dnf install -y rpm-build createrepo_c
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/data/build-repos.sh]: Move this in an prepare install step so that we can run these on the other distros also

File: tests/prepare/artifact/nvr-debug/data/build-repos.sh [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+
+mkdir -p "$REPO_BASE/tc11-high" "$REPO_BASE/tc11-low" \
+         "$REPO_BASE/tc12-a"   "$REPO_BASE/tc12-b"
+
+# Build binary RPMs from the host-built SRPMs for this container's distro.
+rpmbuild --define "_topdir $BUILD_DIR" --rebuild ./srpms/${PACKAGE}-1.0*.src.rpm
+rpmbuild --define "_topdir $BUILD_DIR" --rebuild ./srpms/${PACKAGE}-2.0*.src.rpm
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/data/build-repos.sh]: for loop build

File: tests/prepare/artifact/nvr-debug/data/plan.fmf (hunk 12 above)
Comment (vaibhavdaren) [reply tests/prepare/artifact/nvr-debug/data/plan.fmf]: As dicussed on slack moved the logic to runner.

**Comment by vaibhavdaren** [comment]
From the hacking session : Added the first two test cases from the https://github.com/teemtee/tmt/issues/4546#issuecomment-3908515743
https://artifacts.dev.testing-farm.io/94c23875-cc56-4c77-879b-c2ab3f76f0ae/#work-the-restubje3ckx_plans-provision-container-the-rest_8_default-0. 
The tests should definitely pass now.

File: tests/prepare/artifact/nvr-debug/test.sh [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+        done
+        PACKAGE="dummy-nvr-test"
+        BUILD_DIR="$(mktemp -d)"
+        for srpm in "$SRPM_DIR"/${PACKAGE}-*.src.rpm; do
+            rlRun "rpmbuild --define '_topdir $BUILD_DIR' --rebuild '$srpm'" 0 "Build binary RPM from $srpm"
+        done
+        rlRun "mkdir -p '$SCRIPT_DIR/data/rpms'"
+        rlRun "cp '$BUILD_DIR'/RPMS/noarch/${PACKAGE}-*.rpm '$SCRIPT_DIR/data/rpms/'"
+        rlRun "rm -rf '$BUILD_DIR'"
+        rlRun "pushd data"
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/test.sh]: Instead of doing the build inside the `data`, make a `tmpdir` and copy the data and `pushd` into there. See for example
https://github.com/teemtee/tmt/blob/ebf609c6c38dace4130a97889956e9f6ba67313b/tests/libraries/local/test.sh#L13-L17

File: tests/prepare/artifact/nvr-debug/test.sh [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+# (repository-file) installs the relevant .repo files before each TC runs.
+
+. /usr/share/beakerlib/beakerlib.sh || exit 1
+. ../../../images.sh || exit 1
+. ../lib/common.sh || exit 1
+
+rlJournalStart
+    rlPhaseStartSetup
+        rlRun "PROVISION_HOW=${PROVISION_HOW:-container}"
+        SCRIPT_DIR="$(dirname $0)"
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/test.sh]: `dirname` is unnecessary, we are always assuming this script is run in the cwd. See the `pushd data`

File: tests/prepare/artifact/nvr-debug/test.sh [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+            rlRun "rpmbuild --define '_topdir $BUILD_DIR' --rebuild '$srpm'" 0 "Build binary RPM from $srpm"
+        done
+        rlRun "mkdir -p '$SCRIPT_DIR/data/rpms'"
+        rlRun "cp '$BUILD_DIR'/RPMS/noarch/${PACKAGE}-*.rpm '$SCRIPT_DIR/data/rpms/'"
+        rlRun "rm -rf '$BUILD_DIR'"
+        rlRun "pushd data"
+        rlRun "run=$(mktemp -d)" 0 "Create run directory"
+        DATA_DIR="$(pwd)"
+        setup_distro_environment
+    rlPhaseEnd
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/test.sh]: Please split this in smaller phases. I would say setting up the `mktemp` and `setup_distro_environment` can live here for consistency, but the rpm building should be in a separate `rlPhaseStartTest`

File: tests/prepare/artifact/nvr-debug/test.sh [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+
+rlJournalStart
+    rlPhaseStartSetup
+        rlRun "PROVISION_HOW=${PROVISION_HOW:-container}"
+        SCRIPT_DIR="$(dirname $0)"
+        SRPM_DIR="$SCRIPT_DIR/data/srpms"
+        rlRun "mkdir -p '$SRPM_DIR'"
+        for spec in "$SCRIPT_DIR"/rpm-data/*.spec; do
+            rlRun "sudo mock --buildsrpm --spec '$spec' --sources '$SCRIPT_DIR/rpm-data/' \
+                --resultdir '$SRPM_DIR' --enable-network" 0 "Build SRPM from $spec"
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/test.sh]:
```suggestion
            rlRun "mock --buildsrpm --spec '$spec' --sources '$SCRIPT_DIR/rpm-data/' \
                --resultdir '$SRPM_DIR'" 0 "Build SRPM from $spec"
```
- Heck no `sudo`
- No `enable-network` unless you justify it

File: tests/prepare/artifact/nvr-debug/test.sh [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+        done
+        PACKAGE="dummy-nvr-test"
+        BUILD_DIR="$(mktemp -d)"
+        for srpm in "$SRPM_DIR"/${PACKAGE}-*.src.rpm; do
+            rlRun "rpmbuild --define '_topdir $BUILD_DIR' --rebuild '$srpm'" 0 "Build binary RPM from $srpm"
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/test.sh]: Can also use `mock` here. Not really an issue for this simple test repos though, both `rpmbuild` and `mock` are fine, just be consistent with it. The separate srpm and binary rpm should still be the case because it will make it easier to reuse and cache stuff.

File: tests/prepare/artifact/nvr-debug/data/tc11.fmf [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+    Repos used:
+      - test-high-priority repo (priority=30) provides dummy-nvr-test-1.0
+      - test-low-priority repo  (priority=50) provides dummy-nvr-test-2.0
+
+    Expected result: dummy-nvr-test-1.0 is installed because the lower
+    numeric priority value (30) overrides the higher package version.
+
+test: ./tc11.sh
+require:
+  - dummy-nvr-test
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/data/tc11.fmf]: To not split too much the files (especially when there is relevant documentation associated with it), can you embed the test inside the plan file using `discover/shell` (should be able to add the `require` inside that also.

File: tests/prepare/artifact/nvr-debug/test.sh [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+        setup_distro_environment
+    rlPhaseEnd
+
+    rlPhaseStartTest "TC 1.1 - Repository priority overrides package version"
+        rlRun "tmt run -i $run --scratch -vvv --all \
+            plan --name /plan-tc11 \
+            provision -h $PROVISION_HOW --image $TEST_IMAGE_PREFIX/$image_name \
+            prepare --insert --how artifact \
+                --provide 'repository-file:file://$DATA_DIR/tc11-high.repo' \
+                --provide 'repository-file:file://$DATA_DIR/tc11-low.repo'" \
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/test.sh]: Can't these be embedded in the static fmf file in these cases (the `parsed.path` should account for what a relative path means (can take it relative to fmf tree root for now) but otherwise seems like it would work). Then you can just do a loop of the the whole `rlPahseStartTest` like in
https://github.com/teemtee/tmt/blob/ebf609c6c38dace4130a97889956e9f6ba67313b/tests/core/enabled/test.sh#L10-L21

File: tests/prepare/artifact/nvr-debug/data/build-repos.sh [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+
+set -ex
+
+REPO_BASE="/tmp/nvr-test"
+
+for spec in "$@"; do
+    package="${spec%%:*}"
+    repo="${spec#*:}"
+    mkdir -p "$REPO_BASE/$repo"
+    cp "./rpms/${package}"*.rpm "$REPO_BASE/$repo/"
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/data/build-repos.sh]: This will be problematic in the future when we test same NVR but different contents because the names will clash at build time. Let's make the assumption that all repo content (building and separating in folders) is defined by the layer above in their own folders (although at that point could also do `createrepo` there).

The build steps could be like
```
# Loop over the state of the repos
for repo in repo-1.0 repo-2.0 repo-1.0-broken; do
  # Loop over all spec files in the repos
  pushd $repo
  for spec_file  in *.spec; do
    # for now all of these can build and save the *.rpm in the same cwd
    build_srpm()
    build_rpm()
    create_repo()  # in-place
  done
  popd 
done
```
And inside this script it would just put everything under a predictable file location for the `.repo` files to be able to consume, not even bothering with filtering what is actually used.

File: tests/prepare/artifact/nvr-debug/data/plan-tc11.fmf [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+            rpm -q dummy-nvr-test | grep -E "^dummy-nvr-test-1\.0-"
+            dnf repoquery --installed dummy-nvr-test --qf "%{from_repo}" | grep -Fx "test-high-priority"
+provision:
+    how: container
+prepare:
+    - name: Install createrepo_c
+      order: 10
+      how: install
+      package:
+        - createrepo_c
```
Comment (LecrisUT) [inline tests/prepare/artifact/nvr-debug/data/plan-tc11.fmf]: This can be moved in `main.fmf`

Comment (vaibhavdaren) [reply tests/prepare/artifact/nvr-debug/data/plan-tc11.fmf]: moved the common part to `main.fmf` addressed in https://github.com/teemtee/tmt/pull/4549/changes/9acd9300cb0af28da4410dede3c2c394270715e8.

Comment (LecrisUT) [reply tests/prepare/artifact/nvr-debug/data/plan-tc11.fmf]: You forgot
```yaml
execute:
    how: tmt
```

Comment (vaibhavdaren) [reply tests/prepare/artifact/nvr-debug/data/plan-tc11.fmf]: tq. fixed in https://github.com/teemtee/tmt/pull/4549/commits/1da478bbf381b882d374c11c9944a88f69ea4a00.

File: tests/prepare/artifact/nvr-debug/data/main.fmf [hunk 24]
```diff
@@ -0,0 +1,11 @@
...
+    how: container
+
+execute:
+    how: tmt
+
+prepare:
+    - name: Install createrepo
+      how: install
+      package:
+        - createrepo
```
Comment (tcornell-bus) [inline tests/prepare/artifact/nvr-debug/data/main.fmf]: This prepare step gets overwritten by the prepare step in plan-tc11 and plan-tc12. I think it can be deleted entirely since the tests themselves do not use createrepo and they are currently passing anyway without this step.

Comment (vaibhavdaren) [reply tests/prepare/artifact/nvr-debug/data/main.fmf]: Nice find. Addressed in https://github.com/teemtee/tmt/pull/4549/commits/4179c37559e23a6cc0bfaa88f20f18dbbf21e9c4.

File: tests/prepare/artifact/nvr-debug/data/build-repos.sh [hunk 25]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+# Build local RPM repos from per-repo spec directories.
+# For each subdirectory containing .spec files, builds binary RPMs
+# and runs createrepo_c in-place.
```
Comment (AthreyVinay) [inline tests/prepare/artifact/nvr-debug/data/build-repos.sh]: Possible documentation mis-match? `createrepo_c` vs calling `createrepo .`

Comment (thrix) [reply tests/prepare/artifact/nvr-debug/data/build-repos.sh]: yeah, the package is called `createrepo_c` but it provides `createrepo` executable. IN some older releases it was only `createrepo` package. I would just rename to `createrepo`.

File: tests/prepare/artifact/nvr-priority/main.fmf [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+
+    TC 1.2 - Same Priority favors NVR:
+        When repository priorities are equal, DNF5 falls back to
+        standard RPM version comparison and installs the highest NVR.
+
+link:
+  - verifies: https://github.com/teemtee/tmt/issues/4546
+
+require:
+  - tmt
```
Comment (psss) [inline tests/prepare/artifact/nvr-priority/main.fmf]:
```suggestion
require+:
```

It's a bit better to extend the required packages. In this way we don't lose/override any useful requirements defined in the parent.

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 27]
```diff
@@ -81,8 +83,18 @@ def fetch_contents(
         # It returns an Empty list, as no individual artifact files are downloaded.
 
         self.logger.info(f"Initializing repository provider with URL: {self.id}")
-        # TODO: This should not be using Repository.from_url
-        self.repository = Repository.from_url(url=self.id, logger=self.logger)
+
+        parsed = urlparse(self.id)
+        if parsed.scheme == 'file':
+            # Read .repo file from the local (controller) filesystem.
+            self.logger.info(f"Reading repository file from local path: {parsed.path}")
```
Comment (thrix) [inline tmt/steps/prepare/artifact/providers/repository.py:90]: @vaibhavdaren for debugging purposes, could we also print here the absolute path as a debug output pls?

File: tests/prepare/artifact/nvr-priority/data/build-repos.sh [hunk 28]
```diff
@@ -0,0 +1,19 @@
...
+
+for repo_dir in */; do
+    pushd "$repo_dir"
+    build_dir=$(mktemp -d)
+    for spec in *.spec; do
+        rpmbuild --define "_topdir $build_dir" -bb "$spec"
+    done
+    find "$build_dir/RPMS" -name "*.rpm" -exec cp {} . \;
+    rm -rf "$build_dir"
+    createrepo .
```
Comment (AthreyVinay) [inline tests/prepare/artifact/nvr-priority/data/build-repos.sh:17]: Since you are iterating through all subdirectories which are are not repo directories (eg. `.fmf/`), there is a possibility that this line here will create `repodata/` directories in non repo directories too. Maybe an explicit check to only go over directories that contain `.spec` files?

File: tests/prepare/artifact/nvr-priority/main.fmf
Comment (thrix) [inline tests/prepare/artifact/nvr-priority/main.fmf:1]: Why we call the directory `nvr-debug` could we use some other, more descriptive name please?
Consider renaming to something more descriptive like nvr-priority or repository-priority, matching what the 
tests actually verify (DNF repository priority and NVR resolution behavior).

**Review by thrix — CHANGES_REQUESTED** [review]

File: tests/prepare/artifact/nvr-priority/main.fmf
Comment (vaibhavdaren) [reply tests/prepare/artifact/nvr-priority/main.fmf:1]: I'll change it to `nvr-priority` seems nice. Addressed in f802a19b830c671b96fe584ce6d5b233010f07d4.

File: tests/prepare/artifact/nvr-priority/data/build-repos.sh (hunk 28 above)
Comment (vaibhavdaren) [reply tests/prepare/artifact/nvr-priority/data/build-repos.sh:17]: Thanks for the review! The concern makes sense at first glance, but `*/` in bash                                                                does not match hidden directories (those starting with `.`) unless `shopt -s dotglob` is explicitly set. Since this script only sets `nullglob`,  `.fmf/` will never be matched by the loop.                                                                                                           
                                                                                                                                                       
  In the actual directory layout, `*/` will only ever match the intended repo directories: `tc11-high/`, `tc11-low/`, `tc12-a/`, `tc12-b/` — all of which contain `.spec` files by design, so no stray `repodata/` directories will be created.

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 27 above)
Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py:90]: @thrix : Addressed in f802a19b830c671b96fe584ce6d5b233010f07d4.

File: tests/prepare/artifact/nvr-priority/data/plan-tc12.fmf [hunk 29]
```diff
@@ -0,0 +1,31 @@
...
+              - test-same-priority-b repo (priority=50) provides dummy-nvr-test-2.0
+
+            Expected result: dummy-nvr-test-2.0 is installed because both repos
+            have equal priority and DNF5 falls back to RPM version comparison.
+        require:
+          - dummy-nvr-test
+        test: |
+            set -ex
+            rpm -q dummy-nvr-test | grep -E "^dummy-nvr-test-2\.0-"
+            dnf repoquery --installed dummy-nvr-test --qf "%{from_repo}" | grep -Fx "test-same-priority-b"
```
Comment (psss) [inline tests/prepare/artifact/nvr-priority/data/plan-tc12.fmf]: Bundling the full `tmt` test using the `discover.shell` approach sounds a little bit awkward. What about rather defining a genuine `tmt` test? Both plan and test related to the `tc12` test case could be still defined in the same `tc12.fmf` file:

```yaml
/plan:
    discover:
        how: fmf
    ...

/test:
    summary: TC 1.2 - Same priority falls back to NVR comparison
    description: |
        Verify that when repositories share the same priority, DNF5 selects
        the package with the highest NVR (Name-Version-Release).

        Repos used:
          - test-same-priority-a repo (priority=50) provides dummy-nvr-test-1.0
          - test-same-priority-b repo (priority=50) provides dummy-nvr-test-2.0

        Expected result: dummy-nvr-test-2.0 is installed because both repos
        have equal priority and DNF5 falls back to RPM version comparison.
    require:
      - dummy-nvr-test
    test: |
        set -ex
        rpm -q dummy-nvr-test | grep -E "^dummy-nvr-test-2\.0-"
        dnf repoquery --installed dummy-nvr-test --qf "%{from_repo}" | grep -Fx "test-same-priority-b"
```

**Review by psss — COMMENTED** [review]
Looks good, just two rather minor suggestions.

File: tests/prepare/artifact/nvr-priority/data/plan-tc12.fmf (hunk 29 above)
Comment (vaibhavdaren) [reply tests/prepare/artifact/nvr-priority/data/plan-tc12.fmf]: Nice suggestion looks cleaner this way.  Addressed in e07519aced64d3a8b8f4c19274657b20b56caf58.
